Read only phase1

joshmoore opened the pull request:
Rebase of https://github.com/openmicroscopy/openmicroscopy/pull/5213 onto metadata53

chris-allan committed:
Shim out Event creation into EventProvider

chris-allan committed:
Shim out Node manipulation into NodeProvider

chris-allan committed:
Documentation

chris-allan committed:
Configurable event provider

chris-allan committed:
Configurable node provider

chris-allan committed:
In memory event provider implementation

chris-allan committed:
In memory node provider implemenation

chris-allan committed:
Correctly use interface instead of implementation

chris-allan committed:
Configurable session manager

chris-allan committed:
Also use node provider in session manager

chris-allan committed:
Don't need to double define this bean

chris-allan committed:
Add retrieval by UUID to node provider

chris-allan committed:
When loading a session use the manager

chris-allan committed:
When creating a new event use an unloaded Session

As the Session may be used elsewhere, saving the Event may unload the
Session.  This will have far reaching consequences when the Session
itself came out of the Session cache.

chris-allan committed:
Restore protected visibility

chris-allan committed:
Allow various methods to be overriden by subclasses

chris-allan committed:
Utilise the node provider

chris-allan committed:
Add some more context to the cached session

chris-allan committed:
Refactor session retrieval to allow it to be overriden

chris-allan committed:
Add missing in memory session manager implementation

joshmoore committed:
Fix tests by adding nulls to constructors

joshmoore commented:
Testing:

```
config set -- omero.cluster.node_provider ome.security.basic.InMemoryNodeProvider
config set -- omero.sessions.session_manager ome.services.sessions.InMemorySessionManagerImpl
config set -- omero.security.event_provider ome.security.basic.InMemoryEventProvider
```

~~@chris-allan : looking at this, I'd probably suggest dropping the "Basic" for these implementations.~~

Edit: `Basic` dropped here due to commit below.

joshmoore commented:
Looks on 5.3 at least there's still a code path on start that's causing issues:

```
Caused by: ome.conditions.InternalException:  Wrapped Exception: (org.springframework.jdbc.UncategorizedSQLException):
Hibernate operation: could not insert: [ome.model.enums.Format]; uncategorized SQLException for SQL [insert into format (external_id, p
ermissions, value, id) values (?, ?, ?, ?)]; SQL state [25P02]; error code [0]; ERROR: current transaction is aborted, commands ignored
 until end of transaction block; nested exception is org.postgresql.util.PSQLException: ERROR: current transaction is aborted, commands
 ignored until end of transaction block
        at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.jav
a:83) ~[spring-jdbc.jar:3.0.1.RELEASE]
        at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.jav
a:80) ~[spring-jdbc.jar:3.0.1.RELEASE]
        at org.springframework.orm.hibernate3.HibernateAccessor.convertJdbcAccessException(HibernateAccessor.java:424) ~[spring-orm.jar
:3.0.1.RELEASE]
        at org.springframework.orm.hibernate3.HibernateAccessor.convertHibernateAccessException(HibernateAccessor.java:410) ~[spring-or
m.jar:3.0.1.RELEASE]
        at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:117) ~[spring-orm.jar:3.0.1.RELEASE
]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[spring-aop.jar:3
.0.1.RELEASE]
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108) ~[spring-tx.jar:3
.0.1.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[spring-aop.jar:3
.0.1.RELEASE]
        at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249) ~[server.jar:na]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[spring-aop.jar:3
.0.1.RELEASE]
        at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121) ~[server.jar:na]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[spring-aop.jar:3
.0.1.RELEASE]
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202) ~[spring-aop.jar:3.0.1.RELEASE]
        at com.sun.proxy.$Proxy72.doWork(Unknown Source) ~[na:na]
        at ome.services.util.Executor$Impl.execute(Executor.java:443) ~[server.jar:na]
        at ome.services.util.EnsureEnum.ensure(EnsureEnum.java:100) ~[server.jar:na]
        at ome.services.util.DBEnumCheck.doCheck(DBEnumCheck.java:58) ~[server.jar:na]
        at ome.services.util.BaseDBCheck.start(BaseDBCheck.java:109) ~[server.jar:na]
        at ome.services.util.DBEnumCheck.start(DBEnumCheck.java:31) ~[server.jar:na]
```

joshmoore commented:
I'm going to guess this is a byproduct of the upgrade path. On a new database, it doesn't occur:

```
$ dist/bin/omero -q obj new Project name=foo
Failed to create Project -  Wrapped Exception: (org.springframework.orm.hibernate3.HibernateSystemException):
Exception occurred inside getter of ome.model.meta.Session.annotationLinks; nested exception is org.hibernate.PropertyAccessException: Exception occurred inside getter of ome.model.meta.Session.annotationLinks
```

I'll workaround by starting the server once without the given flags.

joshmoore commented:
Running against a replica produces:

```
2017-05-15 09:50:55,023 WARN  [        ome.services.util.ServiceHandler] (      main) Unknown exception thrown.

org.springframework.jdbc.UncategorizedSQLException: CallableStatementCallback; uncategorized SQLException for SQL [{? = call _prepare_session(?, ?, ?)}]; SQL state [25006]; error code [0]; ERROR: cannot execute CREATE TABLE in a read-only
 transaction
  Where: SQL statement "CREATE TEMP TABLE _current_session ("event_id" int8, "user_id" int8, "group_id" int8) ON COMMIT DELETE ROWS"
PL/pgSQL function _prepare_session(bigint,bigint,bigint) line 5 at SQL statement; nested exception is org.postgresql.util.PSQLException: ERROR: cannot execute CREATE TABLE in a read-only transaction
  Where: SQL statement "CREATE TEMP TABLE _current_session ("event_id" int8, "user_id" int8, "group_id" int8) ON COMMIT DELETE ROWS"
PL/pgSQL function _prepare_session(bigint,bigint,bigint) line 5 at SQL statement
        at ...
        at ome.services.util.Executor$Impl.execute(Executor.java:443) [server.jar:na]
        at ome.services.util.Executor$Impl.execute(Executor.java:387) [server.jar:na]
        at ome.services.scripts.ScriptRepoHelper.loadAllScripts(ScriptRepoHelper.java:457) [server.jar:na]
        at ome.services.scripts.ScriptRepoHelper.loadAll(ScriptRepoHelper.java:522) [server.jar:na]
        at ome.services.scripts.ScriptRepoHelper.loadAll(ScriptRepoHelper.java:543) [server.jar:na]
```

joshmoore commented:
Looks like there's an issue with the repository take over in the read-only state:

```
2017-05-18 20:45:26,250 ERROR [ o.h.e.def.AbstractFlushingEventListener] (2-thread-4) Could not synchronize database state with session
org.hibernate.exception.ConstraintViolationException: could not update: [ome.model.core.OriginalFile#3]
        at org.hibernate.exception.SQLStateConverter.convert(SQLStateConverter.java:96) ~[hibernate-core.jar:na]
...
        at ome.logic.UpdateImpl.saveAndReturnObject(UpdateImpl.java:116) [server.jar:na]
        at ome.services.blitz.repo.AbstractRepositoryI$GetOrCreateRepo.doWork(AbstractRepositoryI.java:405) [blitz.jar:na]
...
        at com.sun.proxy.$Proxy72.doWork(Unknown Source) [na:na]
        at ome.services.util.Executor$Impl.execute(Executor.java:443) [server.jar:na]
        at ome.services.util.Executor$Impl.execute(Executor.java:387) [server.jar:na]
        at ome.services.blitz.repo.AbstractRepositoryI.takeover(AbstractRepositoryI.java:210) [blitz.jar:na]
        at sun.reflect.GeneratedMethodAccessor208.invoke(Unknown Source) ~[na:na]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_131]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_131]
        at org.springframework.util.MethodInvoker.invoke(MethodInvoker.java:273) [spring-core.jar:3.0.1.RELEASE]
        at org.springframework.scheduling.quartz.MethodInvokingJobDetailFactoryBean$MethodInvokingJob.executeInternal(MethodInvokingJobDetailFactoryBean.java:264) [spring-context-support.jar:3.0.1.RELEASE]
...
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
Caused by: org.postgresql.util.PSQLException: ERROR: insert or update on table "originalfile" violates foreign key constraint "fkoriginalfile_update_id_event"
  Detail: Key (update_id)=(-1490) is not present in table "event".
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2270) ~[postgresql.jar:9.4]
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1998) ~[postgresql.jar:9.4]
```

mtbc commented on components/server/src/ome/services/sessions/InMemorySessionManagerImpl.java:
```diff
@@ -0,0 +1,101 @@
+    @Override
+    protected Session findSessionById(Long id, ServiceFactory sf) {
+        for (Session session : sessions.values()) {
+            if (session.getId() == id) {
```
should this use `equals`?

chris-allan replied:
Do you think it should use `equals`? It's comparing two longs and using unboxing. Are you worried about a potential for `null == null`?

mtbc replied:
I was thinking about how ` new Long(1234) !=  new Long(1234)`.

chris-allan replied:
In which case both would be unboxed for the comparison and `new Long(1234) == new Long(1234)` would very much be true. Do you see a reason to be comparing object identity here?

mtbc replied:
```java
final Long x = new Long(1234);
final Long y = new Long(1234);
System.out.println("x == y? " + (x == y));
```
```
x == y? false
```
...is what was bothering me but I may just be missing the obvious.

chris-allan replied:
Weird type coercion due to the unboxing and the print? The assertion `x == y` of the above is definitely true.

```java
final Long x = new Long(1234);
final Long y = new Long(1234);
assert x == y
assert (x == y)
System.out.println("Win!");
```
```
Win!
```

mtbc replied:
Did you try running java with the `-ea` flag to enable assertions?

chris-allan replied:
I'm running in Eclipse where I have assertions enabled by default I believe. Regardless, even more fun:

```java
        final Long x = new Long(1234);
        final Long y = new Long(1234);
        assert x == y;
        assert (x == y);
        System.out.println("x == y? " + new Boolean((x == y)));
        if (x == y) {
            System.out.println("x == y");
        }
        if (x.equals(y)) {
            System.out.println("x.equals(y)");
        }
```
```
x == y? false
x.equals(y)
```

Looks like `equals()` is just plain going to be safer and more explicit.

chris-allan replied:
glencoesoftware/openmicroscopy@8d34d0d

chris-allan replied:
Since this is annoying me I just found the references in the JLS:

 * https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.6.2
 * https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.21.1

`new Long(1234) != new Long(1234)` is definitely always `true`.

In comparisons of `Long` and `Long` object reference is what is compared. If either side of the comparison is `long` unboxing is done and the comparison is by value. So either I don't have assertions enabled like I think I do, I'm trying to do to many things at once, I'm on drugs or all of the above.

`equals()` is most definitely the **only** correct thing to be using in this particular case.

mtbc replied:
Thank you for investigating and digging up the reference!

chris-allan committed:
Do not compare object identity

mtbc commented on components/server/src/ome/security/basic/BasicInMemoryNodeProvider.java:
```diff
@@ -0,0 +1,169 @@
+
+import org.springframework.transaction.annotation.Transactional;
+
+import edu.emory.mathcs.backport.java.util.Collections;
```
The sign of an Eclipse user!

chris-allan replied:
:)

And our mess of a classpath...

joshmoore committed:
Add eventProvider to indexer and pixeldata

joshmoore committed:
Drop 'Basic' from InMemory names

joshmoore committed:
Introduce "ReadOnlyStatus" object

 * Detects that read/write transactions aren't possible
 * and Checks if omero.cluster.read_only is true, then
 * Prevents various r/w actions from taking place.

joshmoore committed:
Catch mysterious NPE on session creation

joshmoore committed:
molecule: read-only test

joshmoore committed:
Simple debugging of incorrect session requests

joshmoore committed:
Debug in-memory session creation

joshmoore committed:
tables: check read_only status

joshmoore committed:
IConfig: methods readOnly by default

joshmoore committed:
Repository: define ROGetOrCreateRepo

Previously, takeover was completely disabled for the service in
order to not start a RW transaction. This splits apart the logic
for the takeover permitting a RO operation which includes reading
from the repo_uuid file and adding the Object to the current
Registry.

joshmoore committed:
FileMaker: propagate readOnly further

FileMaker opened files in "rw" mode. Now the readOnly flag will
switch to "r" and not try to create a .lock file.

fix: filemaker

manics commented:
The server still requires write access to `ManagedRepository/`

joshmoore commented:
Which server tested where? What's the stacktrace?

manics commented:
`OMERO.server-5.3.2-299-5890267-ice36-b75`

```
2017-06-21 21:55:20,801 ERROR [                ome.services.blitz.Entry] (      main) Error on startup.
org.springframework.beans.factory.access.BootstrapException: Unable to return specified BeanFactory instance: factory key [OMERO.blitz], from group with resource name [classpath*:beanRefContext.xml]; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'OMERO.blitz' defined in URL [jar:file:/opt/omero/server/OMERO.server-5.3.2-299-5890267-ice36-b75/lib/server/blitz.jar!/beanRefContext.xml]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [ome.system.OmeroContext]: Constructor threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'managedRepository' defined in URL [jar:file:/opt/omero/server/OMERO.server-5.3.2-299-5890267-ice36-b75/lib/server/blitz.jar!/ome/services/blitz-repositories.xml]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [ome.services.blitz.repo.LegacyRepositoryI]: Constructor threw exception; nested exception is java.lang.IllegalArgumentException: Invalid directory specified for file system service./data/OMERO/ManagedRepository
        at org.springframework.beans.factory.access.SingletonBeanFactoryLocator.useBeanFactory(SingletonBeanFactoryLocator.java:409) ~[spring-beans.jar:3.0.1.RELEASE]
        at ome.system.OmeroContext.getInstance(OmeroContext.java:202) ~[common.jar:na]
        at ome.services.blitz.Entry.start(Entry.java:189) [blitz.jar:na]
        at ome.services.blitz.Entry.main(Entry.java:146) [blitz.jar:na]
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'OMERO.blitz' defined in URL [jar:file:/opt/omero/server/OMERO.server-5.3.2-299-5890267-ice36-b75/lib/server/blitz.jar!/beanRefContext.xml]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [ome.system.OmeroContext]: Constructor threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'managedRepository' defined in URL [jar:file:/opt/omero/server/OMERO.server-5.3.2-299-5890267-ice36-b75/lib/server/blitz.jar!/ome/services/blitz-repositories.xml]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [ome.services.blitz.repo.LegacyRepositoryI]: Constructor threw exception; nested exception is java.lang.IllegalArgumentException: Invalid directory specified for file system service./data/OMERO/ManagedRepository
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:281) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:984) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:888) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:479) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:450) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:290) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:287) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:193) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1048) ~[spring-context.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.access.SingletonBeanFactoryLocator.useBeanFactory(SingletonBeanFactoryLocator.java:397) ~[spring-beans.jar:3.0.1.RELEASE]
        ... 3 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [ome.system.OmeroContext]: Constructor threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'managedRepository' defined in URL [jar:file:/opt/omero/server/OMERO.server-5.3.2-299-5890267-ice36-b75/lib/server/blitz.jar!/ome/services/blitz-repositories.xml]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [ome.services.blitz.repo.LegacyRepositoryI]: Constructor threw exception; nested exception is java.lang.IllegalArgumentException: Invalid directory specified for file system service./data/OMERO/ManagedRepository
        at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:141) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:107) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:273) ~[spring-beans.jar:3.0.1.RELEASE]
        ... 13 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'managedRepository' defined in URL [jar:file:/opt/omero/server/OMERO.server-5.3.2-299-5890267-ice36-b75/lib/server/blitz.jar!/ome/services/blitz-repositories.xml]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [ome.services.blitz.repo.LegacyRepositoryI]: Constructor threw exception; nested exception is java.lang.IllegalArgumentException: Invalid directory specified for file system service./data/OMERO/ManagedRepository
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:281) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:984) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:888) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:479) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:450) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:290) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:287) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:189) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:562) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:871) ~[spring-context.jar:3.0.1.RELEASE]
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:423) ~[spring-context.jar:3.0.1.RELEASE]
        at org.springframework.context.support.ClassPathXmlApplicationContext.<init>(ClassPathXmlApplicationContext.java:139) ~[spring-context.jar:3.0.1.RELEASE]
        at ome.system.OmeroContext.<init>(OmeroContext.java:102) ~[common.jar:na]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_131]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_131]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_131]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_131]
        at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:126) ~[spring-beans.jar:3.0.1.RELEASE]
        ... 15 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [ome.services.blitz.repo.LegacyRepositoryI]: Constructor threw exception; nested exception is java.lang.IllegalArgumentException: Invalid directory specified for file system service./data/OMERO/ManagedRepository
        at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:141) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:107) ~[spring-beans.jar:3.0.1.RELEASE]
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:273) ~[spring-beans.jar:3.0.1.RELEASE]
        ... 33 common frames omitted
Caused by: java.lang.IllegalArgumentException: Invalid directory specified for file system service./data/OMERO/ManagedRepository
        at ome.io.nio.AbstractFileSystemService.<init>(AbstractFileSystemService.java:38) ~[romio.jar:na]
        at ome.io.nio.OriginalFilesService.<init>(OriginalFilesService.java:39) ~[romio.jar:na]
        at ome.services.blitz.repo.LegacyRepositoryI.<init>(LegacyRepositoryI.java:42) ~[blitz.jar:na]
        at ome.services.blitz.repo.LegacyRepositoryI.<init>(LegacyRepositoryI.java:36) ~[blitz.jar:na]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_131]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_131]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_131]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_131]
        at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:126) ~[spring-beans.jar:3.0.1.RELEASE]
        ... 35 common frames omitted
```

manics commented:
I worked around it by symlinking `ManagedRepository/*` as before

joshmoore commented:
Cheers. https://github.com/joshmoore/openmicroscopy/blob/1eb1f66c548c44be9e8b47ab9f2d08b46fcb57ee/components/romio/src/ome/io/nio/AbstractFileSystemService.java#L37

```
    public AbstractFileSystemService(String path) {
        File rootDirectory = new File(path);
        if (!rootDirectory.isDirectory() || !rootDirectory.canRead()
                || !rootDirectory.canWrite()) {
            throw new IllegalArgumentException(
                    "Invalid directory specified for file system service."
		    + rootDirectory);
        }
```

I'll need to inject `readOnlyStatus`

joshmoore commented:
If there are no objections, I'm planning on getting this merged for IDR-0.4.2. It will clearly need more work for the mainline but then we can stop releasing merge builds (as we did albeit not in production for 0.4.0 and 0.4.1)

joshmoore committed:
disable session tests after initial attempt to fix them

joshmoore committed:
Fix flake8

joshmoore commented:
I'm failing to make travis happy this morning. Merging to build this as 0.4.2 and I'll look into getting metadata53 green again.

mtbc commented on components/server/src/ome/security/basic/CurrentDetails.java:
```diff
@@ -302,7 +302,7 @@ public Event newEvent(Session session, EventType type,
         e.setExperimenter(c.getOwner());
         e.setExperimenterGroup(c.getGroup());
-        e.setSession(session);
+        e.setSession(new Session(session.getId(), false));
```
Do you happen to still recall why is it necessary to unload the session at this point?

*Update:* I conjecture that it's to reduce the footprint of `InMemory*` classes.

mtbc commented on components/server/src/ome/security/basic/InMemoryNodeProvider.java:
```diff
@@ -0,0 +1,169 @@
+
+import org.springframework.transaction.annotation.Transactional;
+
+import edu.emory.mathcs.backport.java.util.Collections;
```
May be why Travis failed. *Update:* #5620
